### PR TITLE
GETOK diff/keycloak/visualizer features

### PR DESCRIPTION
**Pull request by loneil**
# Description

CHES access grant:
https://apps.nrs.gov.bc.ca/int/jira/browse/SHOWCASE-245

Text Tweak:
https://apps.nrs.gov.bc.ca/int/jira/browse/SHOWCASE-280

## Types of changes

New feature (non-breaking change which adds functionality) 

 Breaking change (fix or feature that would cause existing functionality to change) 

## Checklist

- [x] I have read the [CONTRIBUTING](CONTRIBUTING.md) doc
- [x] I have checked that unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

## Further comments

Will need in your local json the Keycoak info for each realm (GETOK KC realm, dev, test, prod)

    "keyCloak": {
      "DEV": {
        "endpoint": "SC_KC_INT_ENDPOINT",
        "username": "SC_KC_INT_USERNAME",
        "password": "SC_KC_INT_PASSWORD",
        "realm": "SC_KC_INT_REALM"
      }

### CODE REVIEW

**Comment by bcgov-nr-csst** [comment]
SonarQube Analysis: https://sonarqube-k8vopl-tools.pathfinder.gov.bc.ca/dashboard?id=nr-get-token-pr-86

**Comment by bcgov-nr-csst** [comment]
SonarQube Analysis: https://sonarqube-k8vopl-tools.pathfinder.gov.bc.ca/dashboard?id=nr-get-token-pr-86

**Comment by bcgov-nr-csst** [comment]
SonarQube Analysis: https://sonarqube-k8vopl-tools.pathfinder.gov.bc.ca/dashboard?id=nr-get-token-pr-86

File: frontend/src/components/ConfigForm.vue [hunk 1]
```diff
@@ -135,22 +158,32 @@
...
 
-        <v-btn flat @click="appConfigStep = 2">Back</v-btn>
+        <v-btn flat @click="appConfigStep = 3">Back</v-btn>
 
         <v-dialog
           v-model="confirmationDialog"
           persistent
-          max-width="400"
-          v-if="userAppCfg.deploymentMethod === 'deploymentDirect'"
+          max-width="1200"
```
Comment (jujaga) [inline frontend/src/components/ConfigForm.vue]: Is there any way to make this be reactive and/or be defined as a constants variable elsewhere?

File: frontend/src/components/ConfigForm.vue [hunk 2]
```diff
@@ -135,22 +158,32 @@
...
+              <p>
+                This will overwrite any existing WebADE configuration for the
+                <strong>{{ userAppCfg.applicationAcronym }}</strong> application.
+              </p>
+              <p>The following shows the differences between the current configuration and the configuration that will be created. Are you sure you want to proceed?</p>
+              <pre id="webadeDiff"></pre>
+            </v-card-text>
+            <v-card-text v-if="!usingWebadeConfig">
+              <p>
+                This will create of replace a service client in the Common Services Keycloak realm for the
```
Comment (jujaga) [inline frontend/src/components/ConfigForm.vue]: or, not of

File: frontend/src/components/ConfigForm.vue [hunk 3]
```diff
@@ -299,6 +355,8 @@ import cryptico from 'cryptico-js';
 import Vue from 'vue';
 import VueClipboard from 'vue-clipboard2';
 import { mapGetters } from 'vuex';
+require('colors');
```
[the remark this answers is not in the document]
Comment (loneil) [reply frontend/src/components/ConfigForm.vue]: Removed

File: backend/src/docs/v1.api-spec.yaml [hunk 4]
```diff
@@ -291,7 +291,7 @@ components:
         - applicationDescription
         - commonServices
         - passwordPublicKey
-        - webadeEnvironment
+        - clientEnvironment
```
[the remark this answers is not in the document]
Comment (loneil) [reply backend/src/docs/v1.api-spec.yaml]: Fixed

File: backend/tests/unit/components/realmAdminSvc.spec.js [hunk 5]
```diff
@@ -15,7 +15,7 @@ beforeEach(() => {
     endpoint: realmBaseUrl,
     username: clientId,
     password: clientSecret,
-    realm: realmId} = config.get('serviceClient.keyCloak.INT');
+    realm: realmId} = config.get('serviceClient.keyCloak.DEV');
```
[the remark this answers is not in the document]
Comment (loneil) [reply backend/tests/unit/components/realmAdminSvc.spec.js]: Fixed

File: backend/tests/unit/components/keyCloakServiceClientMgr.spec.js [hunk 6]
```diff
@@ -34,7 +34,7 @@ beforeEach(() => {
     endpoint: realmBaseUrl,
     username: clientId,
     password: clientSecret,
-    realm: realmId} = config.get('serviceClient.keyCloak.INT');
+    realm: realmId} = config.get('serviceClient.keyCloak.DEV');
```
[the remark this answers is not in the document]
Comment (loneil) [reply backend/tests/unit/components/keyCloakServiceClientMgr.spec.js]: Fixed

File: frontend/package.json [hunk 7]
```diff
@@ -19,8 +19,10 @@
   "homepage": "https://github.com/bcgov/nr-get-token",
   "dependencies": {
     "axios": "^0.19.0",
+    "colors": "^1.3.3",
```
[the remark this answers is not in the document]
Comment (loneil) [reply frontend/package.json]: Removed the dependency

File: frontend/src/common/apiService.js [hunk 8]
```diff
@@ -112,5 +113,21 @@ Body: ${JSON.stringify(response.data, null, 2)}`;
...
+  async getWebAdeConfig(webAdeEnv, acronym) {
+    try {
+      const url = `${ApiRoutes.WEBADECONFIG}/${webAdeEnv}/${acronym}`;
+      const response = await apiAxios.get(url);
+
+      return response.data;
+    } catch (e) {
+      // If it's a 404, treat it as a blank config. Not found is valid, it means the acronym doesn't exist
+      if (e.response && e.response.status === 404) {
+        return '';
```
[the remark this answers is not in the document]
Comment (loneil) [reply frontend/src/common/apiService.js]: Will be handled in https://apps.nrs.gov.bc.ca/int/jira/browse/SHOWCASE-308

File: frontend/src/utils/constants.js [hunk 9]
```diff
@@ -3,8 +3,10 @@ const baseRoot = '/api';
 const apiRoot = `${baseRoot}/v1`;
 export const ApiRoutes = Object.freeze({
   APPCONFIG: `${apiRoot}/appConfigForm`,
+  KCCONFIG: `${apiRoot}/kcClientForm`,
   DOCS: `${apiRoot}/docs`,
-  HEALTH: `${apiRoot}/checks/status`
+  HEALTH: `${apiRoot}/checks/status`,
+  WEBADECONFIG: `${apiRoot}/appConfig`
```
[the remark this answers is not in the document]
Comment (loneil) [reply frontend/src/utils/constants.js]: Fixed

File: frontend/src/components/ConfigForm.vue [hunk 10]
```diff
@@ -26,11 +26,28 @@
...
+      <v-btn color="primary" @click="setChes(); appConfigStep = 3" :disabled="!hasAcronyms">
+        <v-icon left>email</v-icon>Common Hosted Email
+      </v-btn>
+      <v-btn color="primary" @click="appConfigStep = 3" :disabled="true">
+        <v-icon left>insert_drive_file</v-icon>Common Hosted Document
+      </v-btn>
+      <v-btn color="primary" @click="setWebade(); appConfigStep = 3" :disabled="!hasAcronyms">
+        <v-icon left>save</v-icon>Legacy (WebADE)
+      </v-btn>
+    </v-stepper-content>
```
[the remark this answers is not in the document]
Comment (loneil) [reply frontend/src/components/ConfigForm.vue]: Noted, will take a look with next round of changes

File: frontend/src/components/ConfigForm.vue (hunk 1 above)
Comment (loneil) [reply frontend/src/components/ConfigForm.vue]: max-width values in html is pretty standard, would set in the CSS if elsewhere

File: frontend/src/components/ConfigForm.vue (hunk 2 above)
Comment (loneil) [reply frontend/src/components/ConfigForm.vue]: Fixed

File: frontend/src/components/ConfigForm.vue [hunk 11]
```diff
@@ -425,6 +492,34 @@ export default {
...
+        acronym: this.userAppCfg.applicationAcronym
+      });
+      var diff = jsdiff.diffLines(
+          this.existingWebAdeConfig,
+          this.appConfigAsString
+        ),
+        display = document.getElementById('webadeDiff'),
+        fragment = document.createDocumentFragment();
+
+      diff.forEach(function(part) {
```
[the remark this answers is not in the document]
Comment (loneil) [reply frontend/src/components/ConfigForm.vue]: Fixed

File: frontend/src/utils/commonServices.js [hunk 12]
```diff
@@ -2,19 +2,25 @@ export default Object.freeze([{
...
   shortName: 'dgen-api',
   name: 'Document Generation Service',
-  apiStoreLink: 'https://i1apistore.nrs.gov.bc.ca/store/apis/info?name=dgen-api&version=v1&provider=admin',
+  apiDocLink: 'https://i1apistore.nrs.gov.bc.ca/store/apis/info?name=dgen-api&version=v1&provider=admin',
   disabled: true
+},
+{
+  abbreviation: 'ches',
+  shortName: 'common-hosted-email',
+  name: 'Common Hosted Email Service',
```
[the remark this answers is not in the document]
Comment (loneil) [reply frontend/src/utils/commonServices.js]: Added a Type field to filter common services on, should no longer appear in the webade list. Thanks!

File: backend/src/docs/v1.api-spec.yaml [hunk 13]
```diff
@@ -362,7 +362,7 @@ components:
             type: string
             example: CHES
           description: An array of applicationAcronyms
-        keycloakEnvironment:
+        clientEnvironment:
```
Comment (jujaga) [inline backend/src/docs/v1.api-spec.yaml]: Reorder this property to be before commonServices (alphabetical ordering)

Comment (loneil) [reply backend/src/docs/v1.api-spec.yaml]: Fixed

**Pull request merged**